driver-core/src/main/java/com/datastax/driver/core/SessionManager.java

```diff
@@ -137,6 +137,13 @@ public String getLoggedKeyspace() {
 
   @Override
   public ResultSetFuture executeAsync(final Statement statement) {
+    if (this.isClosed()) {
+      throw new IllegalStateException("Could not send request, session is closed");
+    }
+
+    if (cluster.isClosed()) {
+      throw new IllegalStateException("Could not send request, cluster is closed");
+    }
```

Comment:
If `!isInit`, we take the second branch below and call `execute()`, which does this again.

Reply:
That was my intuition too, but there is some debate about that see [discussion here](https://github.com/datastax/java-driver/pull/1074#discussion_r208751693).  I need to test it out but i think you are right.

Reply:
I see, the logic at line 144 addresses `RequestHandler(...).sendRequest()` directly, if it called `execute(future, statement)` instead, we wouldn't need this.